### PR TITLE
Support for SUNDIALS v6

### PR DESCRIPTION
**Pull request by cjvogl**
- Added support for SUNDIALS v6, including
  - creation of a `Sundials` singleton class that contains `SUNContext` object introduced in SUNDIALS v6
  - update of all calls to SUNDIALS functions to include `SUNContext` object
  - update of SUNDIALS examples to call `Sundials::Init()` and use new ARKODE enum datatypes and constants
- Refactored global `SundialsMemHelper` object to instead be included as a part of `Sundials` singleton
- Maintained backwards compatibility with SUNDIALS v5 with introduction of deprecated wrappers and datatype definitions
- Removed deprecated functions to encourage use of `SundialsNVector` over raw `N_Vector` and to avoid changes to import approach to support `Sundials` singleton class
  - `HypreParVector::ToNVector`
  - `Vector::Vector(N_Vector)`
  - `Vector::ToNVector`
  
To do:
- [x] update `CHANGELOG` if reviewers are ok with the removal of deprecated functions listed above
- [x] test with CUDA enabled (I could some help with that as I haven't run on any lab GPUs)
- [ ] if preferred by editors, squash commits
- [x] (**Postponed:** fixing this issue seems to require changes in the memory manager which will be done in a separate PR) ~~(added by @v-dobrev) from #2938: *Addition: When addressing this, we should also address an issue seen when running `ex16p` when HYPRE is built with GPU support -- there is a repeated registration of the same host pointer which leads to problems.*~~
- [x] (added by @v-dobrev) from #2938: *Fix an issue: `SUNDIALS_DIR` is set by default to `@MFEM_DIR@/../sundials-5.0.0/instdir` which does not work since we try to expand the path with `$(wildcard ...)` before replacing `@MFEM_DIR@` with the real path.*

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3170](https://github.com/mfem/mfem/pull/3170) | @cjvogl | @pazner | @balos1 + @jandrej + @gardner48 | 8/29/22 | 10/6/22 | 10/13/22 | |

<details>
<summary>PR Checklist</summary>
    
- [x] Code builds.
- [x] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [x] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [x] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [x] Update `INSTALL`:
    - [x] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [x] Have the version ranges for any required or optional libraries changed?
    - [x] Does `make` or `cmake` have a new target?
    - [x] Did the requirements or the installation process change? *(rare)*
- [x] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [x] `.github`
    - [x] `.appveyor.yml`
- [x] Update `.gitignore`:
    - [x] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [x] Add new patterns (just for the new files above) and re-run the above test.
- [x] New examples:
    - [x] All sample runs at the top of the example source file work.
    - [x] Update `examples/makefile`:
      - [x] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [x] Add any files generated by it to the `clean` target.
      - [x] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [x] Update `examples/CMakeLists.txt`:
      - [x] Add the example code to the `ALL_EXE_SRCS` variable.
      - [x] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [x] List the new example in `doc/CodeDocumentation.dox`.
   - [x] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [x] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [x] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [x] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [x] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [x] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [x] New miniapps:
   - [x] All sample runs at the top of the miniapp source file work.
   - [x] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [x] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [x] Update CMake build system:
      - [x] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [x] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [x] Consider adding a new test for the new miniapp.
   - [x] List the new miniapp in `doc/CodeDocumentation.dox`
   - [x] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [x] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [x] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [x] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [x] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [x] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [x] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [x] New capability:
   - [x] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [x] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [x] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [x] Consider adding new sample runs in existing examples to highlight the new capability.
   - [x] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [x] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [x] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [x] Update this checklist, if the new pull request affects it.
- [x] Run `make unittest` to make sure all unit tests pass.
- [x] Run the tests in `tests/scripts`.
- [x] (LLNL only) After merging:
   - [x] Update internal tests to include the new features.
</details>

### CODE REVIEW

File: linalg/sundials.cpp [hunk 1]
```diff
@@ -38,111 +37,242 @@
...
-   ~SundialsMemHelper()
-   {
-      SUNMemoryHelper_Destroy(h);
-   }
+Sundials::~Sundials()
+{
+   SUNContext_Free(&context);
+}
+
+#else
```
Comment (cjvogl) [inline linalg/sundials.cpp:183]:
```suggestion
#else // SUNDIALS_VERSION_MAJOR >= 6
```

File: linalg/sundials.cpp [hunk 2]
```diff
@@ -38,111 +37,242 @@
...
+}
 
-   /// Typecasting to SUNDIALS' SUNMemoryHelper type
-   operator SUNMemoryHelper() const { return h; }
+Sundials::~Sundials()
+{
+   // Do nothing
+}
+
+#endif
```
Comment (cjvogl) [inline linalg/sundials.cpp:195]:
```suggestion
#endif // SUNDIALS_VERSION_MAJOR >= 6
```

File: linalg/sundials.cpp [hunk 3]
```diff
@@ -367,14 +497,15 @@ N_Vector SundialsNVector::MakeNVector(bool use_device)
 #ifdef MFEM_USE_CUDA
    if (use_device)
    {
-      x = N_VNewWithMemHelp_Cuda(0, UseManagedMemory(), sunmemHelper);
+      x = N_VNewWithMemHelp_Cuda(0, UseManagedMemory(), sunmemHelper,
```
Comment (balos1) [inline linalg/sundials.cpp:500]: Should `sunmemHelper` be `Sundials::GetMemHelper()`?

File: linalg/sundials.cpp [hunk 4]
```diff
@@ -397,14 +528,15 @@ N_Vector SundialsNVector::MakeNVector(MPI_Comm comm, bool use_device)
       if (use_device)
       {
          x = N_VMake_MPIPlusX(comm, N_VNewWithMemHelp_Cuda(0, UseManagedMemory(),
-                                                           sunmemHelper));
+                                                           sunmemHelper,
```
Comment (balos1) [inline linalg/sundials.cpp:531]: Should `sunmemHelper` be `Sundials::GetMemHelper()`?

**Review by balos1 — CHANGES_REQUESTED** [review]

File: linalg/sundials.cpp (hunk 3 above)
Comment (balos1) [reply linalg/sundials.cpp:500]: Also, need a `SUNContext` argument at the end.

File: linalg/sundials.cpp (hunk 4 above)
Comment (balos1) [reply linalg/sundials.cpp:531]: Also, need a `SUNContext` argument at the end.

**Comment by balos1** [comment]
@cjvogl  I will try this out with CUDA.

File: linalg/sundials.cpp (hunk 3 above)
Comment (cjvogl) [reply linalg/sundials.cpp]: Should all be addressed!

File: linalg/sundials.cpp (hunk 4 above)
Comment (cjvogl) [reply linalg/sundials.cpp]: Should now be addressed!

**Comment by cjvogl** [comment]
> @cjvogl I will try this out with CUDA.

@balos1, very much appreciated!  I did just push a commit with some of the missing components in the CUDA code you pointed out (and some other small changed need to make `gcc 7.3.0` happy)

**Comment by balos1** [comment]
@cjvogl As per our slack discussion, I have pushed up changes which address some issues that came up when I tested with CUDA on Lassen.

**Review by balos1 — APPROVED** [review]

**Comment by balos1** [comment]
@v-dobrev Do you plan to open a follow-up PR to add HIP support?

**Comment by pazner** [comment]
@cjvogl, would you be able to help out addressing some of these outstanding TODO items?

> * [ ]  update `CHANGELOG` if reviewers are ok with the removal of deprecated functions listed above
> * [ ]  (added by @v-dobrev) from [Add support for SUNDIALS v6 #2938](https://github.com/mfem/mfem/issues/2938): _Addition: When addressing this, we should also address an issue seen when running `ex16p` when HYPRE is built with GPU support -- there is a repeated registration of the same host pointer which leads to problems._
> * [ ]  (added by @v-dobrev) from [Add support for SUNDIALS v6 #2938](https://github.com/mfem/mfem/issues/2938): _Fix an issue: `SUNDIALS_DIR` is set by default to `@MFEM_DIR@/../sundials-5.0.0/instdir` which does not work since we try to expand the path with `$(wildcard ...)` before replacing `@MFEM_DIR@` with the real path._

**Comment by pazner** [comment]
@gardner48, do you think you will be able to take a look at this PR? I think your feedback would be valuable here. We're targeting the MFEM 4.5 release for this feature. No worries if you don't have the time.

**Comment by gardner48** [comment]
@pazner Thanks for the ping, I meant to take a look at this last week. I've got some time this afternoon to review.

**Review by gardner48 — CHANGES_REQUESTED** [review]
One minor typo fix and one obsolete include that can be removed, otherwise LGTM.

**Comment by cjvogl** [comment]
> @cjvogl, would you be able to help out addressing some of these outstanding TODO items?
> 
> > * [ ]   update `CHANGELOG` if reviewers are ok with the removal of deprecated functions listed above
> > * [ ]   (added by @v-dobrev) from [Add support for SUNDIALS v6 #2938](https://github.com/mfem/mfem/issues/2938): _Addition: When addressing this, we should also address an issue seen when running `ex16p` when HYPRE is built with GPU support -- there is a repeated registration of the same host pointer which leads to problems._
> > * [ ]   (added by @v-dobrev) from [Add support for SUNDIALS v6 #2938](https://github.com/mfem/mfem/issues/2938): _Fix an issue: `SUNDIALS_DIR` is set by default to `@MFEM_DIR@/../sundials-5.0.0/instdir` which does not work since we try to expand the path with `$(wildcard ...)` before replacing `@MFEM_DIR@` with the real path._

@pazner, sure thing!  Question for @v-dobrev regarding the `SUNDIALS_DIR`: is this is the `cmake` build system or in the native `makefile` build system... or both?

File: linalg/hypre.cpp [hunk 5]
```diff
@@ -413,15 +413,6 @@ HypreParVector::~HypreParVector()
...
 }
 
-#ifdef MFEM_USE_SUNDIALS
-
-N_Vector HypreParVector::ToNVector()
-{
-   return N_VMake_Parallel(GetComm(), Size(), GlobalSize(), GetData());
-}
-
-#endif // MFEM_USE_SUNDIALS
```
[the remark this answers is not in the document]
Comment (cjvogl) [reply linalg/hypre.cpp:423]: Done!

**Comment by cjvogl** [comment]
@pazner and @v-dobrev, I'm not sure I understand enough about the issue mentioned in 
> - [ ] (added by @v-dobrev) from https://github.com/mfem/mfem/issues/2938: Addition: When addressing this, we should also address an issue seen when running ex16p when HYPRE is built with GPU support -- there is a repeated registration of the same host pointer which leads to problems.
to be able to address it.

**Comment by pazner** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
One small issue that came up in our nightly tests is that we get a number of deprecation warnings since we test with SUNDIALS v5.8.0 and this PR is setup to give these warnings when using SUNDIALS < v6.

Do we want to:
1. keep the deprecation warnings for SUNDIALS < v6 and switch to testing with SUNDIALS v6.3.0, or
2. remove the deprecation warnings for SUNDIALS < v6?

**Comment by tzanio** [comment]
I am okay with both options, with a slight preference for 1️⃣

**Comment by v-dobrev** [comment]
> I am okay with both options, with a slight preference for 1️⃣

I'm not hearing any other preferences so let's go with option 1️⃣ (don't change anything in this PR) which basically means that we should be ready to merge this in `master`.

**Review by jandrej — APPROVED** [review]

**Pull request merged**